### PR TITLE
WAZO-2509 is_running now checks if bus_consumer is actuallly connected

### PR DESCRIPTION
**Pull request by ja0x**
Why: 
  * Previous implementation only checked if thread was alive, now property `is_running` also checks if connection with rabbitmq is alive

### CODE REVIEW

File: xivo_bus/mixins.py [hunk 1]
```diff
@@ -116,35 +116,47 @@ def __init__(self, subscribe=None, **kwargs):
...
 
+    @property
+    @contextmanager
+    def __binding_channel(self):
+        if not self.__connection.connected:
+            self.__connection.connect()
+        yield self.__connection.default_channel
+
     def __create_binding(self, headers, routing_key):
         B = Binding(self.__exchange, routing_key, headers, headers)
```
Comment (fblackburn1) [inline xivo_bus/mixins.py:134]: Please rename this variable to something more verbose :pray:

File: xivo_bus/mixins.py [hunk 2]
```diff
@@ -233,49 +250,82 @@ def __on_message_received(self, body, message):
...
             message.ack()
 
+    def on_connection_error(self, exc, interval):
+        self.log.error(
+            'Broker connection error: %s, trying to reconnect in %s seconds...',
+            exc,
+            interval,
+        )
+        if self.should_stop:
+            raise SystemExit
```
Comment (fblackburn1) [inline xivo_bus/mixins.py:260]: I can understand this usage on Threadable context, but this mixin can be use without thread. Is it safe/good practice do raise SystemExit in *normal* context?

File: xivo_bus/mixins.py [hunk 3]
```diff
@@ -233,49 +250,82 @@ def __on_message_received(self, body, message):
...
+
+    @property
+    def should_stop(self):
+        return getattr(self, 'is_stopping', True)
+
     def on_consume_ready(self, connection, channel, consumers, **kwargs):
         if 'ready_flag' in kwargs:
-            kwargs['ready_flag'].set()
+            ready_flag = kwargs.pop('ready_flag')
+            ready_flag.set()
```
Comment (fblackburn1) [inline xivo_bus/mixins.py:281]: Why we now need to pop the kwargs?

File: xivo_bus/mixins.py [hunk 4]
```diff
@@ -286,22 +336,24 @@ def __init__(self, **kwargs):
...
-                    publish(payload, headers=headers, routing_key=routing_key)
+                    with self.Producer(connection, **publisher_args) as publish:
+                        publish(payload, headers=headers, routing_key=routing_key)
+                except OperationalError as exc:
+                    self.log.error('Publishing queue error: %s', exc, exc_info=1)
+                else:
                     self.__fifo.task_done()
-                except Exception:
-                    self.log.exception('Error while publishing')
+            connection.release()
```
Comment (fblackburn1) [inline xivo_bus/mixins.py:356]: No need to release connection, the `Connection` contextmanager will already release it

**Review by fblackburn1 — CHANGES_REQUESTED** [review]
You talked me about a celery bug/limitation but I don't see any explicit mention? If still valid, can you add explicit comment to help the next developer to make migration to newer celery version

side note: This library will be a pain in the ass to bump to another version --> too much internal celery logic

File: xivo_bus/mixins.py (hunk 2 above)
Comment (fblackburn1) [reply xivo_bus/mixins.py:260]: Always executed in a Thread

File: xivo_bus/mixins.py (hunk 3 above)
Comment (ja0x) [reply xivo_bus/mixins.py:281]: This flag is only useful during initialization to unblock the main thread, this flag serves no other purpose once started and can be removed

**Review by fblackburn1 — APPROVED** [review]

**Pull request merged**